src/Controls/src/Core/Page/Page.cs

```diff
@@ -922,5 +922,123 @@ private protected override string GetDebuggerDisplay()
 			var debugText = DebuggerDisplayHelpers.GetDebugText(nameof(BindingContext), BindingContext, nameof(Title), Title);
 			return $"{this.GetType().FullName}: {debugText}";
 		}
+
+#nullable enable
+		IDisposable? _handleNavigatedEventsForRootPage = null!;
+
+		internal void WireUpAsOutgoingPage(Page? outgoingPage, NavigationType navigationType)
+		{
+			_handleNavigatedEventsForRootPage?.Dispose();
+			_handleNavigatedEventsForRootPage = null;
+
+			if (HasNavigatedTo)
+			{
+				SendNavigatingFrom(new NavigatingFromEventArgs());
+			}
+
+			if (IsLoaded)
+			{
+				_handleNavigatedEventsForRootPage =
+					this.OnUnloaded(() =>
+					{
+						if (HasNavigatedTo)
+						{
+							SendNavigatedFrom(new NavigatedFromEventArgs(outgoingPage, navigationType));
+						}
+
+						outgoingPage?.DisconnectHandlers();
+						_handleNavigatedEventsForRootPage?.Dispose();
+						_handleNavigatedEventsForRootPage = null;
+						outgoingPage = null;
+					});
+			}
+			else
+			{
+				if (HasNavigatedTo)
+				{
+					SendNavigatedFrom(new NavigatedFromEventArgs(outgoingPage, navigationType));
+				}
+
+				outgoingPage?.DisconnectHandlers();
```

Comment:
Shoudl this be set to null?

Maybe to encapsulate this logic and share it, use a local function so we never have to have 2 places to maintain.